bob-apps/bin/bobc.rs

```diff
@@ -1,71 +1,376 @@
-                put(key, size, addr).await;
+    let app_args = AppArgs::from_matches(&matches);
+    info!("{:?}", app_args);
+
+    let addr = NetConfig::from_args(&app_args).get_uri();
+    match app_args.subcommand {
+        PUT_SC => {
+            let keys_names = match app_args.file_pattern {
+                Some(FilePattern::WithRE(re, dir)) => {
+                    // maybe create filenames from pattern AND keys and try to send them?
+                    if app_args.key_pattern != None {
+                        return error!("Either file pattern or key argument must be used");
+                    }
+                    prepare_put_from_pattern(&re.get_regex(), &dir, app_args.keysize).await
+                }
+                Some(FilePattern::WithoutRE(path)) => match app_args.key_pattern {
+                    None => {
+                        return error!("Key arg is required if not using file pattern");
+                    }
+                    Some(p) => match p {
+                        KeyPattern::Single(key) => {
+                            let key = get_key_value(key, app_args.keysize);
+                            vec![KeyName {
+                                key,
+                                name: path.to_string(),
+                            }]
+                        }
+                        _ => return error!("Multiple keys are not allowed without pattern"),
+                    },
+                },
+
+                None => unreachable!(),
+            };
+            for kn in keys_names {
+                info!(
+                    "PUT key: \"{:?}\" to \"{}\" from file \"{}\"",
+                    kn.key, addr, &kn.name
+                );
+                put(kn.key, &kn.name, addr.clone()).await; // i dont like this clone() but BobApiClient::connect requires owning
+            }
+        }
+        GET_SC => {
+            let keys_names = match app_args.file_pattern {
+                Some(FilePattern::WithRE(re, _)) => {
+                    if app_args.key_pattern == None {
+                        return error!("Key arg is required when using pattern");
+                    }
+
+                    let files = re.get_filenames(&app_args.key_pattern.unwrap());
+                    let keys_names = prepare_get(files, app_args.keysize);
+                    info!("{:?}", keys_names);
+                    keys_names
+                }
+                Some(FilePattern::WithoutRE(path)) => match app_args.key_pattern {
+                    Some(KeyPattern::Single(val)) => {
+                        prepare_get(vec![(val, path.to_string())], app_args.keysize)
+                    }
+                    Some(KeyPattern::Range(_)) | Some(KeyPattern::Multiple(_)) => {
+                        return error!("Multiple keys are not allowed without pattern")
+                    }
+                    None => unreachable!(),
```

Comment:
Same here and below - just unwrap (`expect("required value")`) the values in match

Reply:
[80e7fd9](https://github.com/qoollo/bob/pull/562/commits/80e7fd9e39e73d0a86f5496f2132ea3500f9eab6)